Source/EasyNetQ/Internals/TaskHelpers.cs

```diff
@@ -96,54 +125,64 @@ public static class TaskHelpers
                 {
                     tcs.SetException(exception);
                 }
+
                 return tcs.Task;
             };
         }
 
-        public static Task FromCancelled()
+        /// <summary>
+        ///     This is an internal API that supports the EasyNetQ infrastructure and not subject to
+        ///     the same compatibility as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new EasyNetQ release.
+        /// </summary>
+        public static void AttachCancellation<T>(
+            this TaskCompletionSource<T> taskCompletionSource, CancellationToken cancellationToken
+        )
         {
-            var tcs = new TaskCompletionSource<object>();
-            tcs.SetCanceled();
-            return tcs.Task;
-        }
+            if (
+                (taskCompletionSource.Task.CreationOptions & TaskCreationOptions.RunContinuationsAsynchronously) !=
+                TaskCreationOptions.RunContinuationsAsynchronously
+            )
+                throw new ArgumentOutOfRangeException(nameof(taskCompletionSource));
 
-        public static Task<T> FromCancelled<T>()
-        {
-            var tcs = new TaskCompletionSource<T>();
-            tcs.SetCanceled();
-            return tcs.Task;
-        }
+            if (!cancellationToken.CanBeCanceled)
+                return;
 
-        public static Task FromException(Exception exception)
-        {
-            var tcs = new TaskCompletionSource<object>();
-            tcs.SetException(exception);
-            return tcs.Task;
-        }
-
-        public static Task<T> FromException<T>(Exception exception)
-        {
-            var tcs = new TaskCompletionSource<T>();
-            tcs.SetException(exception);
-            return tcs.Task;
+            var state = new TcsWithCancellationToken<T>(taskCompletionSource, cancellationToken);
+            state.CancellationTokenRegistration = cancellationToken.Register(
+                s =>
+                {
+                    var t = (TcsWithCancellationToken<T>) s;
+                    t.Tcs.TrySetCanceled(t.CancellationToken);
+                },
+                state,
+                false
+            );
+            taskCompletionSource.Task.ContinueWith(
+                (_, s) =>
+                {
+                    var r = (TcsWithCancellationToken<T>) s;
```

Comment:
```suggestion
                    var r = (TcsWithCancellationToken<T>)s;
```

Reply:
To be honest, R# by default formats with space. So, til editor config is not fixed, I'm going to use R# default one. Sorry.

Reply:
OK